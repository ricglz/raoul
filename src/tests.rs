use std::{
    fs::{read_dir, ReadDir},
    io::Read,
};

use super::{parse, parse_ast, QuadrupleManager, VM};

fn parse_ast_has_error(filename: &str) {
    println!("Testing {:?}", filename);
    let program = std::fs::read_to_string(filename).expect(filename);
    let debug = false;
    let ast_response = parse(&program, debug);
    assert!(ast_response.is_ok());
    let ast = ast_response.unwrap();
    let res = parse_ast(ast, debug);
    assert!(res.is_err());
    insta::assert_debug_snapshot!(res.unwrap_err());
}

impl<R: Read> VM<R> {
    pub fn new_with_reader(quad_manager: &QuadrupleManager, debug: bool, reader: R) -> Self {
        VM::base_new(quad_manager, debug, Some(reader))
    }
}

fn parse_ast_is_ok(filename: &str) -> QuadrupleManager {
    println!("Testing {:?}", filename);
    let program = std::fs::read_to_string(filename).expect(filename);
    let debug = false;
    let ast_response = parse(&program, debug);
    assert!(ast_response.is_ok());
    let ast = ast_response.unwrap();
    let res = parse_ast(ast, debug);
    assert!(res.is_ok());
    let quad_manager = res.unwrap();
    insta::assert_debug_snapshot!(quad_manager);
    quad_manager
}

fn run_vm_is_error(filename: &str) {
    let quad_manager = parse_ast_is_ok(filename);
    let mut vm = VM::new_with_reader(&quad_manager, false, b"test".as_ref());
    let res = vm.run();
    assert!(res.is_err());
    insta::assert_display_snapshot!(res.unwrap_err());
    insta::assert_debug_snapshot!(vm.messages);
}

fn run_vm_is_ok(filename: &str) {
    let quad_manager = parse_ast_is_ok(filename);
    let mut vm = VM::new_with_reader(&quad_manager, false, b"test".as_ref());
    let res = vm.run();
    assert!(res.is_ok());
    insta::assert_debug_snapshot!(vm.messages);
}

fn expect_paths<F>(paths: ReadDir, mut f: F)
where
    F: FnMut(&str),
{
    for path in paths {
        let file_path = path.expect("File must exist").path();
        let file = file_path.to_str().unwrap();
        f(file);
    }
}

#[test]
fn ast_parsing_invalid_files() {
    let paths = read_dir("examples/invalid/static").unwrap();
    expect_paths(paths, parse_ast_has_error);
}

#[test]
fn vm_running_invalid_files() {
    let paths = read_dir("examples/invalid/dynamic").unwrap();
    expect_paths(paths, run_vm_is_error);
}

#[test]
fn valid_files() {
    let paths = read_dir("examples/valid").unwrap();
    for path in paths {
        let file_path = path.expect("File must exist").path();
        let file = file_path.to_str().unwrap();
        if file == "examples/valid/complete.ra" {
            continue;
        }
        run_vm_is_ok(file);
    }
}
