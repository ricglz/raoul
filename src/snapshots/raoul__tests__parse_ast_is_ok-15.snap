---
source: src/tests.rs
expression: quad_manager
---
QuadrupleManager {
    function_name: "main",
    jump_list: [],
    missing_return: false,
    dir_func: DirFunc {
        functions: {
            "main": Function {
                args: [],
                local_addresses: AddressManager(18, 0, 0, 0),
                name: "main",
                return_type: VOID,
                temp_addresses: TempAddressManager({
                    STRING: [],
                    BOOL: [
                        2750,
                    ],
                    FLOAT: [],
                    INT: [
                        2001,
                        2002,
                    ],
                }),
                variables: {
                    "d": Variable {
                        address: 1014,
                        data_type: INT,
                        dimensions: (
                            Some(
                                2,
                            ),
                            Some(
                                2,
                            ),
                        ),
                        name: "d",
                    },
                    "b": Variable {
                        address: 1004,
                        data_type: INT,
                        dimensions: (
                            Some(
                                2,
                            ),
                            Some(
                                3,
                            ),
                        ),
                        name: "b",
                    },
                    "i": Variable {
                        address: 1003,
                        data_type: INT,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "i",
                    },
                    "a": Variable {
                        address: 1000,
                        data_type: INT,
                        dimensions: (
                            Some(
                                3,
                            ),
                            None,
                        ),
                        name: "a",
                    },
                    "j": Variable {
                        address: 1010,
                        data_type: INT,
                        dimensions: (
                            None,
                            None,
                        ),
                        name: "j",
                    },
                    "c": Variable {
                        address: 1011,
                        data_type: INT,
                        dimensions: (
                            Some(
                                3,
                            ),
                            None,
                        ),
                        name: "c",
                    },
                },
                first_quad: 0,
            },
        },
        global_fn: GlobalScope {
            addresses: AddressManager(0, 0, 0, 0),
            variables: {},
        },
    },
    memory: ConstantMemory({STRING: [Array 'a', Array 'b', Array 'c', Array 'd'], BOOL: [], INT: [1000, 0, 3, 1, 2, 1004, 1011, 1014], FLOAT: []}),
    pointer_memory: PointerMemory {
        counter: 4015,
        pointers: {},
    },
    quad_list: [
        Goto       -     -     1    ,
        Print      3500  -     -    ,
        PrintNl    -     -     -    ,
        Ver        3001  3002  -    ,
        Sum        3000  3001  4000 ,
        Assignment 3003  -     4000 ,
        Ver        3003  3002  -    ,
        Sum        3000  3003  4001 ,
        Assignment 3004  -     4001 ,
        Ver        3004  3002  -    ,
        Sum        3000  3004  4002 ,
        Assignment 3002  -     4002 ,
        Assignment 3001  -     1003 ,
        Lte        1003  3004  2750 ,
        GotoF      2750  -     21   ,
        Ver        1003  3002  -    ,
        Sum        3000  1003  4003 ,
        Print      4003  -     -    ,
        PrintNl    -     -     -    ,
        Inc        -     -     1003 ,
        Goto       -     -     13   ,
        Print      3501  -     -    ,
        PrintNl    -     -     -    ,
        Ver        3001  3004  -    ,
        Times      3001  3002  2000 ,
        Ver        3001  3002  -    ,
        Sum        3005  2000  2001 ,
        Sum        2001  3001  4004 ,
        Assignment 3003  -     4004 ,
        Ver        3001  3004  -    ,
        Times      3001  3002  2001 ,
        Ver        3003  3002  -    ,
        Sum        3005  2001  2002 ,
        Sum        2002  3003  4005 ,
        Assignment 3004  -     4005 ,
        Ver        3001  3004  -    ,
        Times      3001  3002  2002 ,
        Ver        3004  3002  -    ,
        Sum        3005  2002  2001 ,
        Sum        2001  3004  4006 ,
        Assignment 3002  -     4006 ,
        Ver        3003  3004  -    ,
        Times      3003  3002  2001 ,
        Ver        3001  3002  -    ,
        Sum        3005  2001  2002 ,
        Sum        2002  3001  4007 ,
        Assignment 3003  -     4007 ,
        Ver        3003  3004  -    ,
        Times      3003  3002  2002 ,
        Ver        3003  3002  -    ,
        Sum        3005  2002  2001 ,
        Sum        2001  3003  4008 ,
        Assignment 3004  -     4008 ,
        Ver        3003  3004  -    ,
        Times      3003  3002  2001 ,
        Ver        3004  3002  -    ,
        Sum        3005  2001  2002 ,
        Sum        2002  3004  4009 ,
        Assignment 3002  -     4009 ,
        Assignment 3001  -     1003 ,
        Lte        1003  3003  2750 ,
        GotoF      2750  -     76   ,
        Assignment 3001  -     1010 ,
        Lte        1010  3004  2750 ,
        GotoF      2750  -     74   ,
        Ver        1003  3004  -    ,
        Times      1003  3002  2002 ,
        Ver        1010  3002  -    ,
        Sum        3005  2002  2001 ,
        Sum        2001  1010  4010 ,
        Print      4010  -     -    ,
        PrintNl    -     -     -    ,
        Inc        -     -     1010 ,
        Goto       -     -     63   ,
        Inc        -     -     1003 ,
        Goto       -     -     60   ,
        Print      3502  -     -    ,
        PrintNl    -     -     -    ,
        Assignment 3001  -     1003 ,
        Lte        1003  3004  2750 ,
        GotoF      2750  -     90   ,
        Ver        1003  3002  -    ,
        Sum        3006  1003  4011 ,
        Assignment 1003  -     4011 ,
        Ver        1003  3002  -    ,
        Sum        3006  1003  4012 ,
        Print      4012  -     -    ,
        PrintNl    -     -     -    ,
        Inc        -     -     1003 ,
        Goto       -     -     79   ,
        Print      3503  -     -    ,
        PrintNl    -     -     -    ,
        Assignment 3001  -     1003 ,
        Lte        1003  3003  2750 ,
        GotoF      2750  -     117  ,
        Assignment 3001  -     1010 ,
        Lte        1010  3003  2750 ,
        GotoF      2750  -     115  ,
        Ver        1003  3004  -    ,
        Times      1003  3004  2001 ,
        Ver        1010  3004  -    ,
        Sum        3007  2001  2002 ,
        Sum        2002  1010  4013 ,
        Times      1003  3004  2002 ,
        Sum        2002  1010  2001 ,
        Assignment 2001  -     4013 ,
        Ver        1003  3004  -    ,
        Times      1003  3004  2001 ,
        Ver        1010  3004  -    ,
        Sum        3007  2001  2002 ,
        Sum        2002  1010  4014 ,
        Print      4014  -     -    ,
        PrintNl    -     -     -    ,
        Inc        -     -     1010 ,
        Goto       -     -     96   ,
        Inc        -     -     1003 ,
        Goto       -     -     93   ,
        End        -     -     -    ,
    ],
}
